pyanaconda/modules/payloads/payload/flatpak/source.py

```diff
@@ -0,0 +1,433 @@
+#
+# Query and download sources of Flatpak content
+#
+# Copyright (C) 2024 Red Hat, Inc.
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# the GNU General Public License v.2, or (at your option) any later version.
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY expressed or implied, including the implied warranties of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+# Public License for more details.  You should have received a copy of the
+# GNU General Public License along with this program; if not, write to the
+# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
+
+import json
+import os
+from abc import ABC, abstractmethod
+from configparser import ConfigParser, NoSectionError
+from contextlib import contextmanager
+from functools import cached_property
+from typing import Dict, List, Optional, Tuple
+from urllib.parse import urljoin, urlparse
+
+from blivet.arch import get_arch
+
+from pyanaconda.anaconda_loggers import get_module_logger
+from pyanaconda.core.i18n import _
+from pyanaconda.core.util import requests_session
+from pyanaconda.modules.common.structures.payload import RepoConfigurationData
+from pyanaconda.modules.common.task.progress import ProgressReporter
+from pyanaconda.modules.payloads.base.utils import get_downloader_for_repo_configuration
+
+log = get_module_logger(__name__)
+
+__all__ = ["FlatpakSource", "FlatpakStaticSource", "FlatpakRegistrySource", "NoSourceError"]
+
+
+_CONTAINER_ARCH_MAP = {
+    "x86_64": "amd64",
+    "aarch64": "arm64"
+}
+
+
+def _get_container_arch():
+    """Architecture name as used by docker/podman"""
+    arch = get_arch()
+    return _CONTAINER_ARCH_MAP.get(arch, arch)
+
+
+def _canonicalize_flatpak_ref(ref) -> Tuple[Optional[str], str]:
+    """Split off a collection ID, and add architecture if unspecified
+
+    Turn "org.fedoraproject.Stable:app/org.example.Foo//stable" into
+    ("org.fedoraproject.Stable", "app/org.example.Foo/amd64/stable")
+    """
+
+    collection_parts = ref.split(":", 1)
+    if len(collection_parts) == 2:
+        collection = collection_parts[0]
+        ref = collection_parts[1]
+    else:
+        collection = None
+
+    parts = ref.split("/")
+    if len(parts) != 4:
+        raise RuntimeError("Can't parse reference")
+    if parts[2] == "":
+        parts[2] = get_arch()
+
+    return collection, "/".join(parts)
+
+
+class NoSourceError(Exception):
```

Comment:
If this exception could reach UI (other side of DBus) it should be placed in `modules/common/errors/payload.py` and it needs to have `@dbus_error` decorator. The decorator will change the exception to the DBus error and you can create the exception on the UI side.

Also there is already `SourceError` exception which I think could be used here with a details in the message.

If it is used only internally then I would recommend it to move to `payload/flatpak/errors.py` anyway.

Reply:
It can't reach the UI with the current code structure - both places where it could be thrown are caught in FlatpakManager. Currently we log and ignore to try and make things somewhat robust if a user hasn't set things up properly for a DVD/FTP/HTTP install - you at least get everything else installed. But yes, there is some argument for erroring out as well, which could encourage this to be D-Bus error that just throws through. Do you have an opinion on log+ignore or error out?

Reply:
If it is just about printing error than it should be fine to just use `log.error` without breaking the installation completely.

Reply:
I'm changing the `NoSourceError` to existing `SourceSetupError`.